src/services/executionEnvironment.ts

```diff
@@ -0,0 +1,316 @@
+            containerName,
+            builtin_plugin_locations,
+            '*')
+
+          // Copy builtin modules
+          ansibleConfig.module_locations = await this.copyPluginDocFiles(hostCacheBasePath,
+            containerName,
+            ansibleConfig.module_locations,
+            '**/modules')
+        }
+      } catch(error) {
+        this.connection.window.showErrorMessage(
+          `Exception in ExecutionEnvironment service while fetching docs: ${JSON.stringify(error)}`
+        )
+      } finally {
+        this.cleanUpContainer(containerName)
+      }
+    }
+
+    public wrapContainerArgs(command: string): string {
+      const workspaceFolderPath = URI.parse(this.context.workspaceFolder.uri).path
+      const containerCommand: Array<string> = [this._container_engine]
+      containerCommand.push(...['run', '--rm', '--interactive'])
+      containerCommand.push(...['--workdir', workspaceFolderPath])
+
+      containerCommand.push(...['-v', `${workspaceFolderPath}:${workspaceFolderPath}`])
+      if (this._container_engine === 'podman') {
+        // container namespace stuff
+        containerCommand.push('--group-add=root')
+        containerCommand.push('--ipc=host')
+
+        // docker does not support this option
+        containerCommand.push('--quiet')
+      } else {
+        containerCommand.push(`--user=${process.getuid()}`)
+      }
+      containerCommand.push(`--name ansible_language_server_${uuidv4()}`)
+      containerCommand.push(this._container_image)
+      containerCommand.push(command)
+      const generatedCommand = containerCommand.join(' ')
+      this.connection.console.log(`container engine invocation: ${generatedCommand}`)
+      return generatedCommand
+    }
+
+    public cleanUpContainer(containerName: string): void {
+      [`${this._container_engine} stop ${containerName}`, `${this._container_engine} rm ${containerName}`].forEach((command) => {
+        try {
+          child_process.execSync(command, {
+            cwd: URI.parse(this.context.workspaceFolder.uri).path,
+          });
+        } catch(error) {
+          this.connection.console.log(`cleanup command '${command}' failed with error '${error}'`)
+        }
+      })
+    }
+
+    private isPluginInPath(containerName: string, searchPath: string, pluginFolderPath: string): boolean {
+      //const command = this.wrapContainerArgs(`find ${searchPath} -path '${pluginFolderPath}'`)
+      const command = `${this._container_engine} exec -i ${containerName} find ${searchPath} -path '${pluginFolderPath}'`
+      try {
+        this.connection.console.info(`Executing command ${command}`)
+        const result = child_process.execSync(command, {
+          encoding: 'utf-8',
+        }).trim();
+        return result !== ''
+
+      } catch (error) {
+        this.connection.console.error(error)
+        return false
+      }
+    }
+
+    private runContainer(containerName: string): boolean {
```

Comment:
Instead of running a daemon container for copying the data we could use one of the techniques described here: https://stackoverflow.com/questions/25292198/docker-how-can-i-copy-a-file-from-an-image-to-a-host

Reply:
This was the first approach I tried but volume mounting and copying the file takes 5 times more to copy the plugin files as compared to running the image and using the cp command. I have tried with multiple EE and have seen similar results. I have not dig deeper into the reason for the performance hit with the mount approach. That's the reason I refactored later to use `cp` commanf

Reply:
Not the volume approach. This one:
```bash
id=$(docker create image-name)
docker cp $id:path - > local-tar-file
docker rm -v $id
```
or this: `docker cp $(docker create --rm registry.example.com/ansible-base:latest):/home/ansible/.ssh/id_rsa ./hacked_ssh_key`

If we go with that, we will avoid all the code that goes into managing a running container (and false-positive errors that we get with every second `cleanUpContainer` invocation). And the performance hit will be negligible, or we might even have a performance gain (since no container is actually started).

Reply:
Ah, I missed that one. Nice suggestion, I will try it out and update the PR if all is good.

Reply:
After running `ansible-config` command within EE the values of plugins paths returned by ansibleConfig service within EE are
```
ansibleConfig.module_locations
(3) ['/home/runner/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules', '/usr/local/lib/python3.8/site-packages/ansible/modules']
0:'/home/runner/.ansible/plugins/modules'
1:'/usr/share/ansible/plugins/modules'
2:'/usr/local/lib/python3.8/site-packages/ansible/modules'
length:3
```

and 
```
ansibleConfig.collections_paths
(10) ['/home/runner/.ansible/collections', '/usr/share/ansible/collections', '', '/usr/lib64/python38.zip', '/usr/lib64/python3.8', '/usr/lib64/python3.8/lib-dynload', '/usr/local/lib64/python3.8/site-packages', '/usr/local/lib/python3.8/site-packages', '/usr/lib64/python3.8/site-packages', '/usr/lib/python3.8/site-packages']
0:'/home/runner/.ansible/collections'
1:'/usr/share/ansible/collections'
2:''
3:'/usr/lib64/python38.zip'
4:'/usr/lib64/python3.8'
5:'/usr/lib64/python3.8/lib-dynload'
6:'/usr/local/lib64/python3.8/site-packages'
7:'/usr/local/lib/python3.8/site-packages'
8:'/usr/lib64/python3.8/site-packages'
9:'/usr/lib/python3.8/site-packages'
length:10
```

The `isPluginInPath` method identifies if Ansible plugin is present in a given path by executing `find` command within EE and after that use `docker cp` command for only those path that has plugin files present. For the approach you suggested it will require to create container image each time for a given path even if the path doesn't have  Ansible plugins and this approach IMO is not performant due to multiple invocation of run command ad compared to having a container instance running in background.